File: lab_2.java

// Примеры мультисписка, очереди, дека и приоритетной очереди на Java

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Deque;
import java.util.ArrayDeque;
import java.util.PriorityQueue;

public class lab_1_2 {
    public static void exampleMultilist() {
        System.out.println("=== МУЛЬТИСПИСОК (ArrayList<ArrayList<Integer>>) ===");
        ArrayList<ArrayList<Integer>> matrix = new ArrayList<>();

        ArrayList<Integer> row1 = new ArrayList<>();
        row1.add(1);
        row1.add(2);

        ArrayList<Integer> row2 = new ArrayList<>();
        row2.add(3);
        row2.add(4);

        matrix.add(row1);
        matrix.add(row2);

        System.out.println("Матрица: " + matrix);

        System.out.print("Обход элементов: ");
        for (ArrayList<Integer> row : matrix) {
            for (Integer x : row) {
                System.out.print(x + " ");
            }
        }
        System.out.println("\n");
    }

    public static void exampleQueue() {
        System.out.println("=== ОЧЕРЕДЬ (Queue<String>) ===");
        Queue<String> queue = new LinkedList<>();

        queue.add("task1");
        queue.add("task2");
        queue.add("task3");

        while (!queue.isEmpty()) {
            String current = queue.poll(); // извлекает из головы очереди
            System.out.println("Обрабатывается: " + current);
        }
        System.out.println();
    }

    public static void exampleDeque() {
        System.out.println("=== ДЕК (Deque<Integer>) ===");
        Deque<Integer> deque = new ArrayDeque<>();

        deque.addLast(10);   // добавление справа
        deque.addLast(20);
        deque.addFirst(5);   // добавление слева

        System.out.println("Состояние дека: " + deque);

        Integer left = deque.removeFirst();
        Integer right = deque.removeLast();

        System.out.println("Удалён слева: " + left);
        System.out.println("Удалён справа: " + right);
        System.out.println("Итоговое состояние дека: " + deque);
        System.out.println();
    }

    public static void examplePriorityQueue() {
        System.out.println("=== ПРИОРИТЕТНАЯ ОЧЕРЕДЬ (PriorityQueue<Integer>) ===");

        PriorityQueue<Integer> pq = new PriorityQueue<>(); // min-heap

        pq.offer(20);
        pq.offer(5);
        pq.offer(15);
        pq.offer(1);

        while (!pq.isEmpty()) {
            System.out.println("Следующий элемент: " + pq.poll());
        }
        System.out.println();
    }

    public static void main(String[] args) {
        exampleMultilist();
        exampleQueue();
        exampleDeque();
        examplePriorityQueue();
    }
}

File: lab_2.py
# Примеры мультисписка, очереди, дека и приоритетной очереди на Python

from queue import Queue, PriorityQueue
from collections import deque
import heapq


def example_multilist():
    print("=== МУЛЬТИСПИСОК (вложенный список) ===")
    groups = [
        ["Иван", "Петя"],
        ["Маша", "Катя"],
        ["Дима", "Саша"]
    ]
    print("Исходный мультисписок:", groups)

    # Пример обхода мультисписка
    all_names = []
    for group in groups:
        for name in group:
            all_names.append(name)

    print("Все имена в одном списке:", all_names)
    print()


def example_queue():
    print("=== ОЧЕРЕДЬ (Queue) ===")
    q = Queue()

    # Добавление элементов в очередь
    q.put("task1")
    q.put("task2")
    q.put("task3")

    # Извлечение элементов в порядке FIFO
    while not q.empty():
        current = q.get()
        print("Обрабатывается:", current)
    print()


def example_deque():
    print("=== ДЕК (collections.deque) ===")
    d = deque()

    # Добавляем элементы в оба конца
    d.append("справа_1")
    d.append("справа_2")
    d.appendleft("слева_1")

    print("Состояние дека:", d)

    # Удаляем элементы с обоих концов
    left = d.popleft()
    right = d.pop()
    print("Удалён слева:", left)
    print("Удалён справа:", right)
    print("Итоговое состояние дека:", d)
    print()


def example_priority_queue_queue():
    print("=== ПРИОРИТЕТНАЯ ОЧЕРЕДЬ (queue.PriorityQueue) ===")
    pq = PriorityQueue()

    # Чем меньше число, тем выше приоритет
    pq.put((2, "средний приоритет"))
    pq.put((1, "высокий приоритет"))
    pq.put((3, "низкий приоритет"))

    while not pq.empty():
        priority, item = pq.get()
        print(f"priority={priority}, item={item}")
    print()


def example_priority_queue_heapq():
    print("=== ПРИОРИТЕТНАЯ ОЧЕРЕДЬ (heapq) ===")
    customers = []
    # heappush создаёт min-heap
    heapq.heappush(customers, (2, "Harry"))
    heapq.heappush(customers, (3, "Charles"))
    heapq.heappush(customers, (1, "Riya"))
    heapq.heappush(customers, (4, "Stacy"))

    while customers:
        print(heapq.heappop(customers))
    print()


if __name__ == "__main__":
    example_multilist()
    example_queue()
    example_deque()
    example_priority_queue_queue()
    example_priority_queue_heapq()
